From bef9c191a404fe63d2f055f02680b1005e3b4639 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 18:59:41 -0400
Subject: [fs] btrfs: Add datasum mount option

Message-id: <1407179109-13714-95-git-send-email-sandeen@redhat.com>
Patchwork-id: 87006
O-Subject: [RHEL7.1 PATCH 094/422] btrfs: Add datasum mount option.
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Qu Wenruo <quwenruo@cn.fujitsu.com>

commit d399167d88ea53590d6c0850b2d5534cbd21da02
Author: Qu Wenruo <quwenruo@cn.fujitsu.com>
Date:   Mon Jan 6 09:58:32 2014 +0800

    btrfs: Add datasum mount option.

    Add datasum mount option to enable checksum with
    remount option.

    Signed-off-by: Qu Wenruo <quwenruo@cn.fujitsu.com>
    Reviewed-by: David Sterba <dsterba@suse.cz>
    Signed-off-by: Josef Bacik <jbacik@fb.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/Documentation/filesystems/btrfs.txt b/Documentation/filesystems/btrfs.txt
index 78d49ff..ab627e0 100644
--- a/Documentation/filesystems/btrfs.txt
+++ b/Documentation/filesystems/btrfs.txt
@@ -159,8 +159,10 @@ Options with (*) are default options and will not show in the mount options.
  Enable/disable data copy-on-write for newly created files.
  Nodatacow implies nodatasum, and disables all compression.
 
+  datasum(*)
   nodatasum
- Disable data checksumming for newly created files.
+ Enable/disable data checksumming for newly created files.
+ Datasum implies datacow.
 
   notreelog
  Disable the tree logging used for fsync and O_SYNC writes.
diff --git a/fs/btrfs/super.c b/fs/btrfs/super.c
index 6567865..e84e6cb 100644
--- a/fs/btrfs/super.c
+++ b/fs/btrfs/super.c
@@ -325,6 +325,7 @@ enum {
  Opt_check_integrity_print_mask, Opt_fatal_errors, Opt_rescan_uuid_tree,
  Opt_commit_interval, Opt_barrier, Opt_nodefrag, Opt_nodiscard,
  Opt_noenospc_debug, Opt_noflushoncommit, Opt_acl, Opt_datacow,
+ Opt_datasum,
  Opt_err,
 };
 
@@ -334,6 +335,7 @@ static match_table_t tokens = {
  {Opt_subvolid, "subvolid=%s"},
  {Opt_device, "device=%s"},
  {Opt_nodatasum, "nodatasum"},
+ {Opt_datasum, "datasum"},
  {Opt_nodatacow, "nodatacow"},
  {Opt_datacow, "datacow"},
  {Opt_nobarrier, "nobarrier"},
@@ -434,6 +436,14 @@ int btrfs_parse_options(struct btrfs_root *root, char *options)
    btrfs_info(root->fs_info, "setting nodatasum");
    btrfs_set_opt(info->mount_opt, NODATASUM);
    break;
+  case Opt_datasum:
+   if (btrfs_test_opt(root, NODATACOW))
+    btrfs_info(root->fs_info, "setting datasum, datacow enabled");
+   else
+    btrfs_info(root->fs_info, "setting datasum");
+   btrfs_clear_opt(info->mount_opt, NODATACOW);
+   btrfs_clear_opt(info->mount_opt, NODATASUM);
+   break;
   case Opt_nodatacow:
    if (!btrfs_test_opt(root, COMPRESS) ||
     !btrfs_test_opt(root, FORCE_COMPRESS)) {
-- 
1.7.1